evm/rpc/main.py

```diff
@@ -0,0 +1,115 @@
+import json
+import logging
+
+from evm.exceptions import (
+    ValidationError,
+)
+
+from evm.rpc.modules import (
+    Eth,
+    EVM,
+)
+
+REQUIRED_REQUEST_KEYS = (
+    'id',
+    'jsonrpc',
+    'method',
+)
+
+
+def validate_request(request):
+    for key in REQUIRED_REQUEST_KEYS:
+        if key not in request:
+            raise ValueError("request must include the key %r" % key)
+
+
+def generate_response(request):
+    return {
+        'id': request.get('id', -1),
+        'jsonrpc': request.get('jsonrpc', "2.0"),
+    }
+
+
+class RPCServer:
+    '''
+    This "server" accepts json strings requests and returns the appropriate json string response,
+    meeting the protocol for JSON-RPC defined here: https://github.com/ethereum/wiki/wiki/JSON-RPC
+
+    The key entry point for all requests is :meth:`RPCServer.request`, which
+    then proxies to the appropriate method. For example, see
+    :meth:`RPCServer.eth_getBlockByHash`.
+    '''
+
+    module_classes = (
+        Eth,
+        EVM,
+    )
+
+    def __init__(self, chain):
+        self.modules = {}
+        self.chain = chain
+        for M in self.module_classes:
+            self.modules[M.__name__.lower()] = M(chain)
+
+    def _lookup_method(self, rpc_method):
+        method_pieces = rpc_method.split('_')
+
+        if len(method_pieces) != 2:
+            # This check provides a security guarantee: that it's impossible to invoke
+            # a method with an underscore in it. Only public methods on the modules
+            # will be callable by external clients.
+            raise ValueError("Invalid RPC method: %r" % rpc_method)
+        module_name, method_name = method_pieces
+
+        if module_name not in self.modules:
+            raise ValueError("Module unavailable: %r" % module_name)
+        module = self.modules[module_name]
+
+        try:
+            return getattr(module, method_name)
+        except AttributeError:
+            raise ValueError("Method not implemented: %r" % rpc_method)
+
+    def execute(self, request):
+        '''
+        The key entry point for all incoming requests
+        '''
+        response = generate_response(request)
```

Comment:
*mutability nitpick*

The places where `response` is mutated are bugging me.  Mostly that it's only through strict control flow that we don't end up with a response which has both an `"error"` and `"result"` key.

Suggestions to remove it.

```python
# remove mutation
base_response = generate_response(request)
...
if good:
    response = cytoolz.assoc(base_response, 'result', ...)
else:
    response = cytoolz.assoc(base_response, 'error', ...)

# or moving logic into `generate_response`
if good:
    response = generate_response(request, result=...)
else:
    response = generate_response(request, error=...)  
```

Second option should include validation that only one of error/result are provided or a separate response generation functions for error/result.

Reply:
Ok, I can clean up this logic. The issue I run into with validating that only one of error/result are provided is that it's already an error condition if that check fails. I don't want to raise an exception to crash the RPC server, but there's already an error being displayed. The best answer I have is to append an additional error message to the existing one.

Reply:
>  it's already an error condition if that check fails

I think it's ok to have a hard failure that crashes the server if `generate_response` is called with both of those keywords.  You could also just do two separate functions.

```python
# or moving logic into `generate_response`
if good:
    response = generate_success_response(request, result=...)
else:
    response = generate_error_response(request, error=...)  
```